core/sr-primitives/src/weights.rs

```diff
@@ -44,38 +79,44 @@ pub trait WeighData<T> {
 	fn weigh_data(&self, target: T) -> Weight;
 }
 
-/// Default type used as the weight representative in a `#[weight = x]` attribute.
+/// Means of classifying a dispatchable function.
+pub trait ClassifyDispatch<T> {
+	/// Classify the dispatch function based on input data `target` of type `T`.
+	fn classify_dispatch(&self, target: T) -> DispatchClass;
+}
+
+/// Default type used with the `#[weight = x]` attribute in a substrate chain.
 ///
-/// A user may pass in any other type that implements [`Weigh`]. If not, the `Default`
-/// implementation of [`TransactionWeight`] is used.
-pub enum TransactionWeight {
-	/// Basic weight (base, byte).
-	/// The values contained are the base weight and byte weight respectively.
+/// A user may pass in any other type that implements the correct traits. If not, the `Default`
+/// implementation of [`WeightedTransaction`] is used.
+#[derive(Clone, Copy)]
+pub enum WeightedTransaction {
+	/// A fixed-weight transaction. No dependency on state or input.
 	Fixed(Weight),
-	/// Maximum fee. This implies that this transaction _might_ get included but
-	/// no more transaction can be added. This can be done by setting the
-	/// implementation to _maximum block weight_.
-	Max,
-	/// Free. The transaction does not increase the total weight
-	/// (i.e. is not included in weight calculation).
-	Free,
+	/// An operational transaction. Still incurs a weight-fee but typically has a higher priority.
+	Operational(Weight),
```

Comment:
Change to `FixedOperational`, and introduce `MaxOperational` and `FreeOperational`.